tensorboard/program.py

```diff
@@ -407,16 +409,34 @@ def _fix_mime_types(self):
         mimetypes.add_type("font/woff2", ".woff2")
         mimetypes.add_type("text/html", ".html")
 
+    def _make_grpc_provider(self, addr):
+        options = [
+            ("grpc.max_receive_message_length", 1024 * 1024 * 256),
+        ]
+        channel = grpc.insecure_channel(addr, options=options)
+        stub = grpc_provider.make_stub(channel)
+        provider = grpc_provider.GrpcDataProvider(addr, stub)
+        return provider
+
+    def _make_data_provider(self):
+        """Returns `(data_provider, deprecated_multiplexer)`."""
+        grpc_addr = self.flags.grpc_data_provider
+        if grpc_addr:
+            return (self._make_grpc_provider(grpc_addr), None)
```

Comment:
For my own understanding, how does the `grpc_data_provider` know about the logdir, if we don't pass it the CLI flags?

Reply:
The `grpc_data_provider` doesn’t know about the logdir, nor does it need
to. The data location is populated as `grpc://localhost:6801` based on
the address to which it connects. The logdir that the RPC _server_ reads
from is given as a command-line flag to the server, which is a separate
process entirely.

Reply:
Ahh right, the technical design doc even clarifies that the data loading server is in a separate process.  I forgot about that.

In the longer term, is the plan to keep these 2 processes launched separately?  Or will we have the main TB process spawn the RPC serving process, and pass it a subset of the command line flags?

Reply:
In at least the medium term, I plan for them to remain separate
processes (rather than an in-process C extension), but I do plan for
TensorBoard to launch the RPC server. A sibling of `LocalDataIngester`
will probably take care of that.